toolchain/semantics/semantics_node.h

```diff
@@ -365,6 +358,36 @@ class SemanticsNode {
 // may be worth investigating further.
 static_assert(sizeof(SemanticsNode) == 20, "Unexpected SemanticsNode size");
 
+// Provides base support for use of Id types as DenseMap/DenseSet keys.
+// Instantiated below.
+template <typename Id>
+struct SemanticsIdMapInfo {
+  static inline auto getEmptyKey() -> Id {
+    return Id(llvm::DenseMapInfo<int32_t>::getEmptyKey());
+  }
+
+  static inline auto getTombstoneKey() -> Id {
+    return Id(llvm::DenseMapInfo<int32_t>::getTombstoneKey());
+  }
+
+  static auto getHashValue(const Id& val) -> unsigned {
+    return llvm::DenseMapInfo<int32_t>::getHashValue(val.index);
+  }
+
+  static auto isEqual(const Id& lhs, const Id& rhs) -> bool {
+    return lhs == rhs;
+  }
+};
+
 }  // namespace Carbon
 
+// Support use of Id types as DenseMap/DenseSet keys.
+template <>
+struct llvm::DenseMapInfo<Carbon::SemanticsStringId>
+    : public Carbon::SemanticsIdMapInfo<Carbon::SemanticsStringId> {};
+
+template <>
+struct llvm::DenseMapInfo<Carbon::SemanticsNodeId>
+    : public Carbon::SemanticsIdMapInfo<Carbon::SemanticsNodeId> {};
+
```

Comment:
I would have expected these to need to be in the `llvm` namespace? :shrug: not a big deal if Clang is happy.

Reply:
You can blame zygoloid for this one, I mimicked his code:
https://github.com/carbon-language/carbon-lang/blob/trunk/explorer/interpreter/matching_impl_set.h#L120

My expectation here is that because this specifies `struct llvm::` in the name, it's declaring it's overloading that specific template, rather than existing in the local namespace. So we neatly dodge a namespace scope. :)